lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -1905,6 +1919,16 @@ balanceTransactionWithSelectionStrategy
       where
          unUTxO (Cardano.UTxO u) = u
 
+    ptx@(PartialTx partialTx inputUTxO redeemers) =
+        let
+            PartialTx tx u rs =  ptx'
+            ledgerPParams = Cardano.toLedgerPParams
+                    (Cardano.shelleyBasedEra @era)
+                    nodePParams
+            Just recentEra = WriteTx.toRecentEra $ Cardano.cardanoEra @era
+        in
+            PartialTx (increaseZeroAdaOutputs recentEra ledgerPParams tx) u rs
+
```

Comment:
The amount of logic in these `where` clauses is starting to get a bit much, yes. I would like to move balanceTx to a separate module before attempting to splitting out, though.

Reply:
Although it would be enough to move the `increaseZeroAdaOutuputs` to `balanceTransaction` as opposed to `balanceTransactionWithSelectionStrategy`. The only problem is that the `balanceTransactionWithSelectionStrategy` name then makes less sense with both zero-ada and selection strategy differing, so perhaps `balanceTransactionInternal`? 🤨

Reply:
Another option is to allow the caller of `W.balanceTransaction` to _also_ call `W.increaseZeroAdaOutputs` if/as they wish. `W.increaseZeroAdaOutputs` originates from the wallet HTTP API, and _might_ potentially be considered out-of-place or ad-hoc by Marlowe or other users.

The problem here is that we'd currently be unable to remove `guardZeroAdaOutputs`. We'd have to wait for full balanceTx re-use and removal of zero-ada-output-increasing in the underlying coin-selection.

Reply:
Went with 

> Although it would be enough to move the increaseZeroAdaOutuputs to balanceTransaction as opposed to balanceTransactionWithSelectionStrategy. The only problem is that the balanceTransactionWithSelectionStrategy name then makes less sense with both zero-ada and selection strategy differing, so perhaps balanceTransactionInternal?

but with `balanceTransactionWithSelectionStrategyAndNoZeroAdaAdjustment` instead of `balanceTransactionInternal `.